File: frontend/src/lib/api/targets.ts
import { api } from './client';
import type {
	Target,
	TargetCreate,
	TargetUpdate,
	TargetValidationRequest,
	TargetValidationResponse,
	TargetType,
	TargetBulkCreateRequest,
	TargetBulkCreateResponse,
	TargetImportRequest,
	EnrichmentRefreshResponse
} from '$lib/types/target';

import type { TargetDetailRead } from '@/types/target-detail';

import type { PaginatedResponse, TargetCounts } from '$lib/types/pagination';
import type { SignalFilter, SortDir, SortKey, TargetSummary } from '$lib/utilities/target-signals';

interface ListTargetsParams {
	project_slug?: string;
	search?: string;
	organization_ids?: string[];
	tag_ids?: string[];
	target_type?: TargetType;
	signal?: SignalFilter | null;
	sort_by?: SortKey;
	sort_dir?: SortDir;
	page?: number;
	size?: number;
}

type TargetStatsParams = Pick<
	ListTargetsParams,
	'project_slug' | 'search' | 'organization_ids' | 'tag_ids' | 'target_type'
>;

function buildTargetQuery(params: ListTargetsParams | TargetStatsParams): URLSearchParams {
	const sp = new URLSearchParams();
	if (params.project_slug) sp.append('project_slug', params.project_slug);
	if ('search' in params && params.search?.trim()) sp.append('search', params.search.trim());
	for (const id of params.organization_ids ?? []) sp.append('organization_ids', id);
	for (const id of params.tag_ids ?? []) sp.append('tag_ids', id);
	if (params.target_type) sp.append('target_type', params.target_type);
	return sp;
}

export const targetsApi = {
	async list(params?: ListTargetsParams): Promise<PaginatedResponse<Target>> {
		const sp = buildTargetQuery(params ?? {});
		if (params?.signal) sp.append('signal', params.signal);
		if (params?.sort_by) sp.append('sort_by', params.sort_by);
		if (params?.sort_dir) sp.append('sort_dir', params.sort_dir);
		if (params?.page) sp.append('page', params.page.toString());
		if (params?.size) sp.append('size', params.size.toString());

		const query = sp.toString();
		return api.get<PaginatedResponse<Target>>(query ? `/targets?${query}` : '/targets');
	},

	async getStats(params: TargetStatsParams): Promise<TargetSummary> {
		const query = buildTargetQuery(params).toString();
		return api.get<TargetSummary>(`/targets/stats?${query}`);
	},

	async getMatchingIds(params: ListTargetsParams): Promise<string[]> {
		const sp = buildTargetQuery(params);
		if (params.signal) sp.append('signal', params.signal);
		return api.get<string[]>(`/targets/ids?${sp.toString()}`);
	},

	async bulkEnrich(
		targetIds: string[],
		kind: 'whois' | 'dns' | 'bgp'
	): Promise<{ queued: number }> {
		return api.post<{ queued: number }>('/targets/enrich/bulk', {
			target_ids: targetIds,
			kind
		});
	},

	async bulkAddTags(targetIds: string[], tagNames: string[]): Promise<{ updated: number }> {
		return api.post<{ updated: number }>('/targets/tags/bulk', {
			target_ids: targetIds,
			tag_names: tagNames
		});
	},

	async bulkAddOrganizations(
		targetIds: string[],
		organizationNames: string[]
	): Promise<{ updated: number }> {
		return api.post<{ updated: number }>('/targets/organizations/bulk', {
			target_ids: targetIds,
			organization_names: organizationNames
		});
	},

	async getCounts(projectSlug: string): Promise<TargetCounts> {
		return api.get<TargetCounts>(`/targets/counts?project_slug=${projectSlug}`);
	},

	async get(targetId: string): Promise<Target> {
		return api.get<Target>(`/targets/${targetId}`);
	},

	async create(data: TargetCreate): Promise<Target> {
		return api.post<Target>('/targets', data);
	},

	async update(targetId: string, data: TargetUpdate): Promise<Target> {
		return api.patch<Target>(`/targets/${targetId}`, data);
	},

	async delete(targetId: string): Promise<void> {
		return api.delete(`/targets/${targetId}`);
	},

	async validate(data: TargetValidationRequest): Promise<TargetValidationResponse> {
		return api.post<TargetValidationResponse>('/targets/validate', data);
	},

	async bulkCreate(data: TargetBulkCreateRequest): Promise<TargetBulkCreateResponse> {
		return api.post<TargetBulkCreateResponse>('/targets/bulk', data);
	},

	async importJson(data: TargetImportRequest): Promise<TargetBulkCreateResponse> {
		return api.post<TargetBulkCreateResponse>('/targets/import/json', data);
	},

	async importCsv(projectSlug: string, file: File): Promise<TargetBulkCreateResponse> {
		const formData = new FormData();
		formData.append('file', file);

		const response = await fetch(`/api/v1/targets/import/csv?project_slug=${projectSlug}`, {
			method: 'POST',
			body: formData,
			credentials: 'include'
		});

		if (!response.ok) {
			const errorData = await response.json().catch(() => ({}));
			throw new Error(errorData.detail || `API Error: ${response.status}`);
		}

		return response.json();
	},

	async refreshDns(targetId: string): Promise<EnrichmentRefreshResponse> {
		return api.post<EnrichmentRefreshResponse>(`/targets/${targetId}/dns/refresh`);
	},

	async refreshWhois(targetId: string): Promise<EnrichmentRefreshResponse> {
		return api.post<EnrichmentRefreshResponse>(`/targets/${targetId}/whois/refresh`);
	},

	async refreshBgp(targetId: string): Promise<EnrichmentRefreshResponse> {
		return api.post<EnrichmentRefreshResponse>(`/targets/${targetId}/bgp/refresh`);
	},

	async getDetail(targetId: string): Promise<TargetDetailRead> {
		return api.get<TargetDetailRead>(`/targets/${targetId}/detail`);
	}
};

File: api/app/api/v1/targets.py
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile, status
from fastapi_pagination import Page
from fastapi_pagination.ext.sqlalchemy import paginate
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUser
from app.core.database import get_session
from app.services.target import TargetService
from app.services.target_filters import SignalName, SortDir, SortKey
from shared.definitions.constants import MAX_TARGET_IMPORT
from shared.models import (
    TargetBulkCreate,
    TargetBulkCreateResponse,
    TargetCreate,
    TargetImportRequest,
    TargetRead,
    TargetType,
    TargetUpdate,
    TargetValidationRequest,
    TargetValidationResponse,
)
from shared.schemas.target_detail import (
    EnrichmentRefreshResponse,
    TargetBgpDetailResponse,
    TargetDetailRead,
    TargetDnsDetailResponse,
    TargetWhoisDetailResponse,
)

router = APIRouter(
    prefix="/targets",
    tags=["targets"],
)


def get_target_service(
    session: Annotated[AsyncSession, Depends(get_session)],
) -> TargetService:
    return TargetService(session)


@router.post("/validate", response_model=TargetValidationResponse)
async def validate_target_endpoint(
    request: TargetValidationRequest,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    target_type = await service.validate_target_value(request.target_value)

    if target_type:
        return TargetValidationResponse(
            valid=True,
            target_type=target_type,
            error=None,
            target_value=request.target_value,
        )

    return TargetValidationResponse(
        valid=False,
        target_type=None,
        error="Invalid target format. Accepted formats: domain/subdomain, IP, IP range (CIDR), ASN (AS followed by numbers), or URL",
        target_value=request.target_value,
    )


@router.post("/validate/bulk", response_model=list[TargetValidationResponse])
async def validate_bulk_target(
    request: list[TargetValidationRequest],
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    if len(request) > MAX_TARGET_IMPORT:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Maximum {MAX_TARGET_IMPORT} targets allowed per request",
        )

    seen = set()
    unique_requests = []
    for req in request:
        if req.target_value not in seen:
            seen.add(req.target_value)
            unique_requests.append(req)

    results = []
    for req in unique_requests:
        target_type = await service.validate_target_value(req.target_value)

        if target_type:
            results.append(
                TargetValidationResponse(
                    valid=True,
                    target_type=target_type,
                    error=None,
                    target_value=req.target_value,
                )
            )
        else:
            results.append(
                TargetValidationResponse(
                    valid=False,
                    target_type=None,
                    error="Invalid target format. Accepted formats: domain/subdomain, IP, IP range (CIDR), ASN (AS followed by numbers), or URL",
                    target_value=req.target_value,
                )
            )

    return results


@router.get("/counts", response_model=dict[str, int])
async def get_target_counts(
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
    project_slug: Annotated[str, Query(description="Filter by project slug")],
):
    return await service.get_target_counts(project_slug)


class TargetStatsResponse(BaseModel):
    total: int
    expiring: int
    attention: int
    awaiting: int
    enriched: int


@router.get("/stats", response_model=TargetStatsResponse)
async def get_target_stats(
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
    project_slug: Annotated[str, Query(description="Filter by project slug")],
    search: Annotated[str | None, Query(description="Search target value/name")] = None,
    organization_ids: Annotated[
        list[UUID] | None, Query(description="Filter by organization IDs")
    ] = None,
    tag_ids: Annotated[
        list[UUID] | None, Query(description="Filter by tag IDs")
    ] = None,
    target_type: Annotated[
        TargetType | None, Query(description="Filter by target type")
    ] = None,
):
    return await service.get_target_stats(
        project_slug=project_slug,
        search=search,
        organization_ids=organization_ids,
        tag_ids=tag_ids,
        target_type=target_type,
    )


@router.get("/ids", response_model=list[UUID])
async def list_matching_target_ids(
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
    project_slug: Annotated[str, Query(description="Filter by project slug")],
    search: Annotated[str | None, Query()] = None,
    organization_ids: Annotated[list[UUID] | None, Query()] = None,
    tag_ids: Annotated[list[UUID] | None, Query()] = None,
    target_type: Annotated[TargetType | None, Query()] = None,
    signal: Annotated[SignalName | None, Query()] = None,
):
    return await service.get_matching_target_ids(
        project_slug=project_slug,
        search=search,
        organization_ids=organization_ids,
        tag_ids=tag_ids,
        target_type=target_type,
        signal=signal,
    )


MAX_BULK_TARGET_IDS = 10000
MAX_BULK_NAMES = 100


class BulkEnrichRequest(BaseModel):
    target_ids: list[UUID] = Field(..., min_length=1, max_length=MAX_BULK_TARGET_IDS)
    kind: Literal["whois", "dns", "bgp"]


class BulkTagRequest(BaseModel):
    target_ids: list[UUID] = Field(..., min_length=1, max_length=MAX_BULK_TARGET_IDS)
    tag_names: list[str] = Field(..., min_length=1, max_length=MAX_BULK_NAMES)


class BulkOrgRequest(BaseModel):
    target_ids: list[UUID] = Field(..., min_length=1, max_length=MAX_BULK_TARGET_IDS)
    organization_names: list[str] = Field(..., min_length=1, max_length=MAX_BULK_NAMES)


@router.post("/enrich/bulk", response_model=dict[str, int])
async def bulk_enrich_targets(
    request: BulkEnrichRequest,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    queued = await service.bulk_enrich(request.target_ids, request.kind)
    return {"queued": queued}


@router.post("/tags/bulk", response_model=dict[str, int])
async def bulk_add_tags(
    request: BulkTagRequest,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    updated = await service.bulk_add_tags(
        request.target_ids, request.tag_names, current_user.id
    )
    return {"updated": updated}


@router.post("/organizations/bulk", response_model=dict[str, int])
async def bulk_add_organizations(
    request: BulkOrgRequest,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    updated = await service.bulk_add_organizations(
        request.target_ids, request.organization_names, current_user.id
    )
    return {"updated": updated}


@router.get("/search", response_model=Page[TargetRead])
async def search_targets_by_value(
    _current_user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
    service: Annotated[TargetService, Depends(get_target_service)],
    target_value: Annotated[str, Query(description="Target value to search for")],
    project_slug: Annotated[
        str | None, Query(description="Optional: Filter by specific project")
    ] = None,
):
    query = await service.search_targets_by_value(
        target_value=target_value,
        project_slug=project_slug,
    )

    return await paginate(
        session,
        query,
        transformer=lambda items: [service._to_target_read(t) for t in items],
    )


@router.get("", response_model=Page[TargetRead])
async def list_targets(
    _current_user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
    service: Annotated[TargetService, Depends(get_target_service)],
    project_slug: Annotated[
        str | None, Query(description="Filter by project slug")
    ] = None,
    search: Annotated[
        str | None, Query(description="Search by target value or display name")
    ] = None,
    organization_ids: Annotated[
        list[UUID] | None, Query(description="Filter by organization IDs")
    ] = None,
    tag_ids: Annotated[
        list[UUID] | None, Query(description="Filter by tag IDs")
    ] = None,
    target_type: Annotated[
        TargetType | None, Query(description="Filter by target type")
    ] = None,
    signal: Annotated[
        SignalName | None,
        Query(description="Filter by signal: expiring, attention, awaiting, enriched"),
    ] = None,
    sort_by: Annotated[SortKey, Query(description="Sort field")] = "updated",
    sort_dir: Annotated[SortDir, Query(description="Sort direction")] = "desc",
):
    query = await service.list_targets(
        project_slug=project_slug,
        search=search,
        organization_ids=organization_ids,
        tag_ids=tag_ids,
        target_type=target_type,
        signal=signal,
        sort_by=sort_by,
        sort_dir=sort_dir,
    )

    return await paginate(
        session,
        query,
        transformer=lambda items: [service._to_target_read(t) for t in items],
    )


@router.post("", response_model=TargetRead, status_code=status.HTTP_201_CREATED)
async def create_target(
    target_in: TargetCreate,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.create_target(target_in, current_user.id)


@router.post(
    "/bulk",
    response_model=TargetBulkCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
async def bulk_create_targets(
    bulk_in: TargetBulkCreate,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.bulk_create_targets(bulk_in, current_user.id)


@router.post(
    "/import/json",
    response_model=TargetBulkCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
async def import_targets_json(
    import_request: TargetImportRequest,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.import_targets_structured(import_request, current_user.id)


@router.post(
    "/import/csv",
    response_model=TargetBulkCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
async def import_targets_csv(
    project_slug: Annotated[
        str, Query(description="Project slug to import targets into")
    ],
    file: Annotated[UploadFile, File(...)],
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.import_targets_csv(project_slug, file, current_user.id)


@router.get("/{target_id}/detail", response_model=TargetDetailRead)
async def get_target_detail(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.get_target_detail(target_id)


@router.get("/{target_id}/dns", response_model=TargetDnsDetailResponse)
async def get_target_dns(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.get_target_dns(target_id)


@router.get("/{target_id}/whois", response_model=TargetWhoisDetailResponse)
async def get_target_whois(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.get_target_whois(target_id)


@router.get("/{target_id}/bgp", response_model=TargetBgpDetailResponse)
async def get_target_bgp(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.get_target_bgp(target_id)


@router.post("/{target_id}/dns/refresh", response_model=EnrichmentRefreshResponse)
async def refresh_target_dns(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.refresh_target_dns(target_id)


@router.post("/{target_id}/whois/refresh", response_model=EnrichmentRefreshResponse)
async def refresh_target_whois(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.refresh_target_whois(target_id)


@router.post("/{target_id}/bgp/refresh", response_model=EnrichmentRefreshResponse)
async def refresh_target_bgp(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.refresh_target_bgp(target_id)


@router.get("/{target_id}", response_model=TargetRead)
async def get_target(
    target_id: str,
    _current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.get_target(target_id)


@router.patch("/{target_id}", response_model=TargetRead)
async def update_target(
    target_id: str,
    target_in: TargetUpdate,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    return await service.update_target(target_id, target_in, current_user.id)


@router.delete("/{target_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_target(
    target_id: str,
    current_user: CurrentUser,
    service: Annotated[TargetService, Depends(get_target_service)],
):
    await service.delete_target(target_id, current_user.id)
